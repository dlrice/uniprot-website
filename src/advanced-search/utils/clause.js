import { v1 } from 'uuid';
import { serializableDeepAreEqual, removeProperty } from '../../utils/utils';

export const createEmptyClause = () => ({
  id: v1(),
  logicOperator: 'AND',
  field: {
    label: 'Any',
    term: 'All',
    example: 'a4_human, P05067, cdc7 human',
    itemType: 'single',
    dataType: 'string',
    id: 'id_Any',
  },
  queryInput: {},
});

// export const isClauseTouched = (clause) => {
//   const empty = createEmptyClause();
//   const emptyWithoutID = removeProperty(empty, 'id');
//   const clauseWithoutID = removeProperty(clause, 'id');
//   return !serializableDeepEquals(emptyWithoutID, clauseWithoutID);
// };

export const clausesAreEqual = (clause1, clause2) => serializableDeepAreEqual(removeProperty(clause1, 'id'), removeProperty(clause2, 'id'));
