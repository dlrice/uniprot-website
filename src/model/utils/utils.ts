/* eslint-disable @typescript-eslint/no-explicit-any */
const hasContent = (obj: any) => {
  return Object.values(obj).some(val => {
    if (Array.isArray(val)) {
      const valArray = val as any[];
      return valArray.length > 0;
    }
    if (typeof val === 'object' && val) {
      return Object.values(val).length > 0;
    }
    return false;
  });
};

export default hasContent;
