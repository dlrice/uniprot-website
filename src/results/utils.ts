import queryStringModule from 'query-string';
import { SelectedFacet, SortDirection } from './types/resultsTypes';
import { SortableColumn } from '../model/types/ColumnTypes';

const facetsAsArray = (facetString: string): SelectedFacet[] => {
  return facetString.split(',').map((stringItem) => {
    const [name, value] = stringItem.split(':');
    return {
      name,
      value,
    };
  });
};

export const getURLParams = (
  url: string
): {
  query: string;
  selectedFacets: SelectedFacet[];
  sortColumn: SortableColumn;
  sortDirection: SortDirection;
} => {
  const urlParams = queryStringModule.parse(url);
  const { query, facets, sort, dir } = urlParams;

  let selectedFacets: SelectedFacet[] = [];
  if (facets && typeof facets === 'string') {
    selectedFacets = facetsAsArray(facets);
  }
  const sortDirection = dir as keyof typeof SortDirection;

  return {
    query: query && typeof query === 'string' ? query : '',
    selectedFacets,
    sortColumn: sort as SortableColumn,
    sortDirection: sortDirection && SortDirection[sortDirection],
  };
};

export const setUrlParams = () => null;

// const setURLParams = (
//   query: string,
//   selectedFacets: SelectedFacet[],
//   sortColumn?: string,
//   sortDirection?: SortDirection
// ): void => {
//   const { history } = this.props;
//   history.push({
//     pathname: '/uniprotkb',
//     search: [
//       `query=${query}${this.facetsAsString(selectedFacets)}`,
//       `${sortColumn ? `&sort=${sortColumn}` : ''}`,
//       `${sortDirection ? `&dir=${sortDirection}` : ''}`,
//     ].join(''),
//   });
// };
