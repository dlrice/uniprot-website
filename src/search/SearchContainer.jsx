import React, { Component, Fragment } from 'react';
import { connect } from 'react-redux';
import { withRouter } from 'react-router-dom';
import { MainSearch } from 'franklin-sites';
import {
  selectField,
  updateInputValue,
  updateEvidence,
  updateRangeValue,
  updateLogicOperator,
  submitAdvancedQuery,
  addClause,
  removeClause,
  fetchSearchTerms,
  fetchEvidencesIfNeeded,
  updateClauses,
} from './state/actions';
import AdvancedSearch from './AdvancedSearch';
import createQueryString from './utils/QueryStringGenerator';

import './styles/SearchContainer.scss';

export class Search extends Component {
  constructor(props) {
    super(props);
    const { queryString } = props;
    this.state = {
      showAdvanced: false,
      queryString,
    };
    this.toggleAdvanced = this.toggleAdvanced.bind(this);
    this.handleSubmitClick = this.handleSubmitClick.bind(this);
    this.handleAdvancedSubmitClick = this.handleAdvancedSubmitClick.bind(this);
    this.hundleQueryStringChange = this.hundleQueryStringChange.bind(this);
  }

  componentDidMount() {
    const { dispatchFetchSearchTerms } = this.props;
    dispatchFetchSearchTerms();
  }

  componentDidUpdate(prevProps) {
    const { queryString: prevQueryString } = prevProps;
    const { queryString } = this.props;
    if (prevQueryString !== queryString) {
      this.setState({ queryString });
    }
  }

  toggleAdvanced() {
    const { showAdvanced } = this.state;
    this.setState({ showAdvanced: !showAdvanced });
  }

  handleAdvancedSubmitClick() {
    const { history, clauses } = this.props;
    const encodedQueryString = encodeURI(createQueryString(clauses));
    history.push({ pathname: '/uniprotkb', search: `query=${encodedQueryString}` });
  }

  handleSubmitClick(e) {
    e.preventDefault();
    const { history } = this.props;
    const { queryString } = this.state;
    const encodedQueryString = encodeURI(queryString);
    console.log('encodedQueryString', encodedQueryString);
    history.push({ pathname: '/uniprotkb', search: `query=${encodedQueryString}` });
  }

  hundleQueryStringChange(queryString) {
    this.setState({ queryString });
  }

  render() {
    const { showAdvanced, queryString } = this.state;
    let search;
    if (showAdvanced) {
      search = (
        <AdvancedSearch
          {...this.props}
          queryString={queryString}
          handleAdvancedSubmitClick={this.handleAdvancedSubmitClick}
        />
      );
    } else {
      search = (
        <MainSearch
          onSubmit={this.handleSubmitClick}
          onChange={this.hundleQueryStringChange}
          searchTerm={queryString}
        />
      );
    }
    return (
      <Fragment>
        {search}
        <button type="button" onClick={this.toggleAdvanced} className="adv-search-toggle">
          {showAdvanced ? 'Back to quick search' : 'Advanced search'}
        </button>
      </Fragment>
    );
  }
}

const mapStateToProps = state => ({
  clauses: state.query.clauses,
  searchTerms: state.query.searchTerms.data,
  namespace: state.query.namespace,
  evidences: state.query.evidences,
  queryString: state.results.queryString,
});

const mapDispatchToProps = dispatch => ({
  handleFieldSelect: (clauseId, field) => dispatch(selectField(clauseId, field)),
  handleInputChange: (clauseId, value) => dispatch(updateInputValue(clauseId, value)),
  handleEvidenceChange: (clauseId, value) => dispatch(updateEvidence(clauseId, value)),
  handleRangeInputChange: (clauseId, value, from) => dispatch(updateRangeValue(clauseId, value, from)),
  handleLogicChange: (clauseId, value) => dispatch(updateLogicOperator(clauseId, value)),
  handleRemoveClause: clauseId => dispatch(removeClause(clauseId)),
  dispatchAddClause: () => dispatch(addClause()),
  dispatchfetchEvidencesIfNeeded: evidencesType => dispatch(fetchEvidencesIfNeeded(evidencesType)),
  dispatchFetchSearchTerms: () => dispatch(fetchSearchTerms()),
  dispatchSubmitAdvancedQuery: () => dispatch(submitAdvancedQuery()),
  dispatchUpdateClauses: clauses => dispatch(updateClauses(clauses)),
});

const SearchContainer = withRouter(
  connect(
    mapStateToProps,
    mapDispatchToProps,
  )(Search),
);

export default SearchContainer;
