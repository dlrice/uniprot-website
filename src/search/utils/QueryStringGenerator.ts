import { Clause } from '../types/searchTypes';

type IPrefixMap = {
  feature: string;
  comment: string;
  [key: string]: string;
};

const getItemTypePrefix = (itemType: string) => {
  const itemTypeToPrefixMap: IPrefixMap = {
    feature: 'ft_',
    comment: 'cc_',
  };
  return itemTypeToPrefixMap[itemType] || '';
};

const getItemTypeEvidencePrefix = (itemType: string) => {
  const itemTypeToEvidencePrefixMap: IPrefixMap = {
    feature: 'ftev_',
    comment: 'ccev_',
  };
  return itemTypeToEvidencePrefixMap[itemType] || '';
};

const getItemTypeRangePrefix = (itemType: string) =>
  itemType === 'feature' ? 'ftlen_' : '';

const createTermString = (
  term: string | undefined,
  itemType: string,
  id: string | undefined
) => {
  if (term === undefined) {
    throw new Error('term is undefined');
  }
  if (term === 'ec') {
    if (id) {
      return 'ec:';
    }
    throw new Error('ID value not provided in query');
  }
  if (['organism', 'taxonomy', 'host'].includes(term)) {
    if (id) {
      return `${term}_id:`;
    }
    return `${term}_name:`;
  }
  const itemTypePrefix = getItemTypePrefix(itemType);
  return `${itemTypePrefix}${term}${term ? ':' : ''}`;
};

const createValueString = (
  term: string | undefined,
  valuePrefix: string | undefined,
  stringValue: string,
  id: string | undefined
) => {
  if (term === undefined) {
    throw new Error('term is undefined');
  }
  if (term === 'ec') {
    if (id) {
      return id;
    }
    throw new Error('ID value not provided in query');
  }
  if (id) {
    if (['cofactor_chebi', 'catalytic_activity'].includes(term)) {
      return `"${id}"`;
    }
    if (
      ['organism', 'taxonomy', 'host', 'keyword', 'scl_term'].includes(term)
    ) {
      return id;
    }
  }

  // We are testing for term=xref and valuePrefix=any because the
  // search API expects the valuePrefix to be ommited in this case.
  // eg xref:foo rather than xref_any:foo
  const valuePrefixChecked =
    valuePrefix && !(term === 'xref' && valuePrefix === 'any')
      ? `${valuePrefix}-`
      : '';
  return `${valuePrefixChecked}${stringValue}`;
};

const createSimpleSubquery = (clause: Clause) => {
  if (clause.searchTerm.itemType === 'group') {
    throw Error('Cannot create a query with a group term.');
  }
  const { itemType, term, valuePrefix } = clause.searchTerm;
  const { stringValue, id } = clause.queryInput;
  if (!stringValue) {
    throw new Error('Value not provided in query');
  }
  if (term === 'All') {
    return stringValue;
  }
  const termString = createTermString(term, itemType, id);
  const valueString = createValueString(term, valuePrefix, stringValue, id);
  return `(${termString}${valueString})`;
};

const createRangeSubquery = (clause: Clause) => {
  const { term, itemType } = clause.searchTerm;
  const { rangeFrom, rangeTo } = clause.queryInput;
  const rangeFromChecked = rangeFrom || '';
  const rangeToChecked = rangeTo || '';
  const itemTypeRangePrefix = getItemTypeRangePrefix(itemType);
  return `(${itemTypeRangePrefix}${term}:[${rangeFromChecked} TO ${rangeToChecked}])`;
};

const getEvidenceSubquery = (clause: Clause) => {
  const { evidenceValue } = clause.queryInput;
  const { term, itemType } = clause.searchTerm;
  if (!evidenceValue) {
    throw new Error('Evidence value not provided');
  }
  const itemTypeEvidencePrefix = getItemTypeEvidencePrefix(itemType);
  return `(${itemTypeEvidencePrefix}${term}:${evidenceValue})`;
};

const createQueryString = (clauses: Clause[] = []): string =>
  clauses.reduce((queryAccumulator: string, clause: Clause) => {
    const query = [];
    if (
      (clause.queryInput.id && clause.queryInput.id !== '') ||
      (clause.queryInput.stringValue && clause.queryInput.stringValue !== '')
    ) {
      query.push(createSimpleSubquery(clause));
    }
    if (clause.queryInput.rangeFrom || clause.queryInput.rangeTo) {
      query.push(createRangeSubquery(clause));
    }
    if (
      clause.queryInput.evidenceValue &&
      clause.queryInput.evidenceValue !== ''
    ) {
      query.push(getEvidenceSubquery(clause));
    }
    let queryJoined = query.join(' AND ');
    if (query.length > 1) {
      queryJoined = `(${queryJoined})`;
    }

    let logicOperator = '';
    if (queryAccumulator.length > 0 && query.length > 0) {
      logicOperator = ` ${clause.logicOperator} `;
    } else if (
      queryAccumulator.length === 0 &&
      clause.logicOperator === 'NOT'
    ) {
      logicOperator = `${clause.logicOperator} `;
    }

    return `${queryAccumulator}${logicOperator}${queryJoined}`;
  }, '');

export default createQueryString;
