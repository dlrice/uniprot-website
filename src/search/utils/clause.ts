import { v1 } from 'uuid';
import { serializableDeepAreEqual, removeProperty } from '../../utils/utils';
import { Operator, Input, SearchTermType } from '../types/searchTypes';

enum itemType {
  comment = 'comment',
  database = 'database',
  feature = 'feature',
  goterm = 'goterm',
  group = 'group',
  groupDisplay = 'groupDisplay',
  single = 'single',
}

enum dataType {
  empty = '',
  date = 'date',
  enum = 'enum',
  integer = 'integer',
  string = 'string',
}

type Clause = {
  id: string;
  logicOperator: Operator;
  searchTerm: {
    label: string;
    term: string;
    example: string;
    itemType: itemType;
    dataType: dataType;
    id: string;
  };
  queryInput: Input;
};

export const createEmptyClause = (): Clause => ({
  id: v1(),
  logicOperator: Operator.AND,
  searchTerm: {
    label: 'Any',
    term: 'All',
    example: 'a4_human, P05067, cdc7 human',
    itemType: itemType.single,
    dataType: dataType.string,
    id: 'id_Any',
  },
  queryInput: {},
});

export const clausesAreEqual = (clause1: Clause, clause2: Clause) =>
  serializableDeepAreEqual(
    removeProperty(clause1, 'id'),
    removeProperty(clause2, 'id')
  );

const findSearchTermRecursive = (
  queryField: string,
  searchTerms: SearchTermType[],
  valuePrefix = ''
): SearchTermType | undefined => {
  for (const searchTerm of searchTerms) {
    if (
      valuePrefix &&
      searchTerm.valuePrefix === valuePrefix &&
      searchTerm.term === queryField
    ) {
      return searchTerm;
    }
    if (!valuePrefix && searchTerm.term === queryField) {
      return searchTerm;
    }
    if (searchTerm.items) {
      const found = findSearchTermRecursive(
        queryField,
        searchTerm.items,
        valuePrefix
      );
      if (found) {
        return found;
      }
    }
  }
};

const findSearchTerm = (
  queryField: string,
  searchTerms: SearchTermType[],
  valuePrefix = ''
): SearchTermType | undefined => {
  const searchTerm = findSearchTermRecursive(
    queryField,
    searchTerms,
    valuePrefix
  );
  if (!searchTerm) {
    throw new Error(`${queryField} not a valid field.`);
  }
  return searchTerm;
};

const parseRangeValue = (value: string) => {
  const regeNumberRange = /\[([0-9]+)\s*TO\s*([0-9]+)\]/gi;
  const captureGroupsNumber = regeNumberRange.exec(value);
  if (captureGroupsNumber && captureGroupsNumber[1] && captureGroupsNumber[2]) {
    return {
      rangeFrom: Number(captureGroupsNumber[1]),
      rangeTo: Number(captureGroupsNumber[2]),
    };
  }
  const regexDateRange = /\[([0-9\-]+)\s*TO\s*([0-9\-]+)\]/gi;
  const captureGroupsDate = regexDateRange.exec(value);
  if (captureGroupsDate && captureGroupsDate[1] && captureGroupsDate[2]) {
    return {
      rangeFrom: captureGroupsDate[1],
      rangeTo: captureGroupsDate[2],
    };
  }
  throw new Error(`${value} value is not a valid range`);
};

const parseXrefSubquery = (
  term: string,
  value: string,
  conjunction: keyof typeof Operator,
  searchTerms: SearchTermType[]
) => {
  const clause = {
    id: v1(),
    logicOperator: conjunction,
  };
  const tokens = value.split('-');
  if (tokens.length === 1) {
    // Only a value is provided so this is an xref any search
    return {
      ...clause,
      searchTerm: findSearchTerm('xref', searchTerms, 'any'),
      queryInput: { stringValue: tokens[0] },
    } as Clause;
  }
  if (tokens.length === 2) {
    return {
      ...clause,
      searchTerm: findSearchTerm('xref', searchTerms, tokens[0]),
      queryInput: { stringValue: tokens[1] },
    } as Clause;
  }
  throw new Error(`${value} is not a properly formed xref.`);
};

const parseIdNameSubquery = (
  queryTerm: string,
  value: string,
  conjunction: keyof typeof Operator,
  searchTerms: SearchTermType[]
) => {
  const [termName, searchType] = queryTerm.split('_');
  const searchTerm = findSearchTerm(termName, searchTerms);

  if (searchType === 'id') {
    return {
      id: v1(),
      logicOperator: conjunction,
      searchTerm,
      queryInput: { stringValue: value },
    } as Clause;
  }
  throw new Error(`${queryTerm} not a valid _id or _name style term.`);
};

const parseRangeOrEvidenceSubquery = (subquery: string, value: string) => {
  const regexp = /^(?:ft|cc)(len|ev)?_(.*):.*/i;
  const matches = regexp.exec(subquery);
  if (!matches) {
    return {};
  }
  const rangeOrEvidence = matches[1];
  if (!rangeOrEvidence) {
    return { stringValue: value };
  }
  if (rangeOrEvidence === 'ev') {
    return { evidenceValue: value };
  }
  if (rangeOrEvidence === 'len') {
    return { ...parseRangeValue(value) };
  }
  throw new Error(`${subquery} subquery is not valid.`);
};

const parseSinglePartSubquery = (
  conjunctionString: string = 'AND',
  termValue: string,
  searchTerms: SearchTermType[]
) => {
  if (!searchTerms) {
    return;
  }
  const conjunctionUpper = conjunctionString.toUpperCase();

  let conjunction: keyof typeof Operator;
  if (conjunctionUpper in Operator) {
    conjunction = conjunctionUpper as keyof typeof Operator;
  } else {
    throw new Error(`${conjunctionString} conjunction is not valid.`);
  }

  const tokens = termValue.split(':');
  const value = tokens[1];
  let term = tokens[0];
  if (!value) {
    const emptyClause = createEmptyClause();
    return { ...emptyClause, queryInput: { stringValue: term } };
  }

  if (term.match(/^xref/i)) {
    return parseXrefSubquery(term, value, conjunction, searchTerms);
  }

  if (term.match(/^organism|taxonomy/i)) {
    return parseIdNameSubquery(term, value, conjunction, searchTerms);
  }

  let queryInput = {};
  if (term.match(/^ft(len|ev)?_/i)) {
    try {
      queryInput = parseRangeOrEvidenceSubquery(termValue, value);
    } catch (e) {
      // console.log(e);
    }
    term = getSinglePartSubqueryTerm(termValue);
  }

  const searchTerm = findSearchTerm(term, searchTerms);
  if (!searchTerm) {
    throw new Error(`${term} not a valid term.`);
  }

  return {
    id: v1(),
    logicOperator: conjunction,
    searchTerm,
    queryInput: {
      ...queryInput,
      ...(searchTerm.hasRange
        ? parseRangeValue(value)
        : { stringValue: value }),
    },
  } as Clause;
};

interface IPrefixMap {
  ft: string;
  cc: string;
  [key: string]: string;
}

const itemTypeToPrefixMap: IPrefixMap = {
  ft: 'feature',
  cc: 'comment',
};

const getSinglePartSubqueryPrefix = (subquery: string) => {
  const prefix = subquery.slice(0, 2);
  return prefix in itemTypeToPrefixMap ? prefix : null;
};

const allStringArrayElementsEqual = (arr: Array<string | null>) => {
  return arr.every(
    (element: string | null, i: number) => !!element && element === arr[0]
  );
};

const getTermFrom = (queryStringGroup: string[]) => {
  const prefixes = queryStringGroup.map((queryString: string) =>
    queryString.slice(0, 2)
  );
  if (prefixes.every((prefix: string, i: number) => prefix === prefixes[0])) {
    return itemTypeToPrefixMap[prefixes[0]];
  } else {
    throw new Error(`${queryStringGroup} has inconsistent item type prefixes.`);
  }
};

const getSinglePartSubqueryTerm = (field: string) => {
  if (field.indexOf('_') < 0) {
    return field;
  }
  const regexp = /^\w*?_(\w+)/g;
  const match = regexp.exec(field);
  if (match && match[1]) {
    return match[1];
  }
  throw new Error(`${field} does not contain a searchtype.`);
};

const getMultiplePartSubqueryTerm = (subqueryParts: string[]) => {
  const terms = subqueryParts.map(subqueryPart =>
    getSinglePartSubqueryTerm(subqueryPart)
  );
  if (allStringArrayElementsEqual(terms)) {
    return terms[0];
  }
};

const getMultiplePartSubqueryPrefix = (subqueryParts: string[]) => {
  const prefixes = subqueryParts.map(subqueryPart =>
    getSinglePartSubqueryPrefix(subqueryPart)
  );
  if (allStringArrayElementsEqual(prefixes)) {
    return prefixes[0];
  }
};

const parseMultiplePartSubquery = (
  conjunction: string,
  subqueryParts: string[],
  searchTerms: SearchTermType[]
) => {
  const term = getMultiplePartSubqueryTerm(subqueryParts);
  if (!term) {
    throw new Error(`${subqueryParts} has inconsistent terms.`);
  }
  const searchTerm = findSearchTerm(term, searchTerms);
  if (!searchTerm) {
    throw new Error(`Cannot find ${term}: in search_terms.`);
  }
  const prefix = getMultiplePartSubqueryPrefix(subqueryParts);
  if (!prefix) {
    throw new Error(`${subqueryParts} has inconsistent prefixes.`);
  }
  let queryInput = {};
  subqueryParts.forEach((subqueryPart: string) => {
    const [field, value] = subqueryPart.split(':');
    queryInput = {
      ...queryInput,
      ...parseRangeOrEvidenceSubquery(subqueryPart, value),
    };
  });

  return {
    id: v1(),
    logicOperator: conjunction,
    searchTerm,
    queryInput,
  };
};

const parseSubquery = (
  conjunction: string,
  subquery: string,
  searchTerms: SearchTermType[]
) => {
  const regexQuery = /\(?([0-9a-z_]+:?[0-9a-z\[\]\s\-_\.]*)\)?(?:\s*(AND|OR|NOT)\s*)*/gi;
  const subqueryParts = [];
  let match;
  do {
    // Need this so we don't assign within a loop: while ((match = regexQuery.exec(subquery)) !== null)
    match = regexQuery.exec(subquery);
    if (match) {
      const fieldValue = match[1];
      subqueryParts.push(fieldValue);
    }
  } while (match);

  switch (subqueryParts.length) {
    case 0:
      throw new Error(`No clauses found in ${subquery}.`);
    case 1:
      return parseSinglePartSubquery(
        conjunction,
        subqueryParts[0],
        searchTerms
      );
    default:
      return parseMultiplePartSubquery(conjunction, subqueryParts, searchTerms);
  }
};

const getTopLevelConjunctionsClauses = (query: string) => {
  let balance = 0;
  const clauses = [];
  const conjunctions = [];
  let start = 0;
  for (let i = 0; i < query.length; i += 1) {
    const character = query[i];
    if (character === '(') {
      balance += 1;
      if (balance === 1) {
        if (i === 0) {
          conjunctions.push('AND');
        } else {
          conjunctions.push(query.slice(start + 1, i).trim());
        }
        start = i;
      }
    } else if (character === ')') {
      balance -= 1;
      if (balance === 0) {
        clauses.push(query.slice(start + 1, i));
        start = i;
      }
    }
  }
  return [conjunctions, clauses];
};

export const parseQueryString = (
  query: string,
  searchTerms: SearchTermType[]
) => {
  if (!query) {
    return [createEmptyClause()];
  }
  const [
    topLevelConjunctions,
    topLevelSubqueries,
  ] = getTopLevelConjunctionsClauses(query);
  const clauses = [];
  for (let i = 0; i < topLevelConjunctions.length; i += 1) {
    const conjuction = topLevelConjunctions[i];
    const clause = topLevelSubqueries[i];
    clauses.push(parseSubquery(conjuction, clause, searchTerms));
  }
  return clauses;
};
