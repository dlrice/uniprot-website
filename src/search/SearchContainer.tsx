import React, { Component, Fragment, SyntheticEvent } from 'react';
import { connect } from 'react-redux';
import { Dispatch, bindActionCreators } from 'redux';
import { withRouter, RouteComponentProps } from 'react-router-dom';
import { MainSearch } from 'franklin-sites';
import { RootState, RootAction } from '../state/state-types';
import * as searchActions from './state/actions';
import {
  Clause, FieldType, Operator, EvidenceType,
} from './types/searchTypes';
import AdvancedSearch from './AdvancedSearch';
import { createQueryString } from './utils/QueryStringGenerator';

import './styles/SearchContainer.scss';

interface SearchProps extends RouteComponentProps {
  queryString: string;
  // dispatchFetchSearchTerms: () => void;
  // dispatchfetchEvidencesIfNeeded: (type: string) => void;
  // dispatchSubmitAdvancedQuery: () => void;
  clauses?: Array<Clause>;
  dispatchUpdateQueryString: (type: string) => void;
  // dispatchAddClause: () => void;
}

interface SearchContainerState {
  showAdvanced: boolean;
  queryString: string;
}

export class Search extends Component<SearchProps, SearchContainerState> {
  constructor(props: SearchProps) {
    super(props);
    const { queryString } = props;
    this.state = {
      showAdvanced: false,
      queryString,
    };
    this.toggleAdvanced = this.toggleAdvanced.bind(this);
    this.handleSubmitClick = this.handleSubmitClick.bind(this);
    this.handleAdvancedSubmitClick = this.handleAdvancedSubmitClick.bind(this);
    this.handleQueryStringChange = this.handleQueryStringChange.bind(this);
  }

  componentDidUpdate(prevProps: SearchProps) {
    const { queryString: prevQueryString } = prevProps;
    const { queryString } = this.props;
    if (prevQueryString !== queryString) {
      this.setState({ queryString });
    }
  }

  toggleAdvanced() {
    const { showAdvanced } = this.state;
    this.setState({ showAdvanced: !showAdvanced });
  }

  handleAdvancedSubmitClick() {
    const { history, clauses, dispatchUpdateQueryString } = this.props;
    const queryString = createQueryString(clauses);
    dispatchUpdateQueryString(queryString);
    history.push({ pathname: '/uniprotkb', search: `query=${queryString}` });
  }

  handleSubmitClick(e: SyntheticEvent) {
    e.preventDefault();
    const { history, dispatchUpdateQueryString } = this.props;
    const { queryString } = this.state;
    dispatchUpdateQueryString(queryString);
    history.push({ pathname: '/uniprotkb', search: `query=${queryString}` });
  }

  handleQueryStringChange(queryString: string) {
    this.setState({ queryString });
  }

  render() {
    const { showAdvanced, queryString } = this.state;
    let search;
    if (showAdvanced) {
      search = (
        <AdvancedSearch
          {...this.props}
          queryString={queryString}
          handleAdvancedSubmitClick={this.handleAdvancedSubmitClick}
        />
      );
    } else {
      search = (
        <MainSearch
          onSubmit={this.handleSubmitClick}
          onChange={this.handleQueryStringChange}
          searchTerm={queryString}
        />
      );
    }
    return (
      <Fragment>
        {search}
        <button type="button" onClick={this.toggleAdvanced} className="adv-search-toggle">
          {showAdvanced ? 'Back to quick search' : 'Advanced search'}
        </button>
      </Fragment>
    );
  }
}

const mapStateToProps = (state: RootState) => ({
  clauses: state.query.clauses,
  searchTerms: state.query.searchTerms.data,
  namespace: state.query.namespace,
  evidences: state.query.evidences,
  queryString: state.results.queryString,
});

const mapDispatchToProps = (dispatch: Dispatch<RootAction>) => bindActionCreators(
  {
    handleFieldSelect: (clauseId: string, field: FieldType) => searchActions.selectField(clauseId, field),
    handleInputChange: (clauseId: string, value: string) => searchActions.updateInputValue(clauseId, value),
    handleEvidenceChange: (clauseId: string, value: string) => searchActions.updateEvidence(clauseId, value),
    handleRangeInputChange: (clauseId: string, value: number, from: boolean) => searchActions.updateRangeValue(clauseId, value, from),
    handleLogicChange: (clauseId: string, value: Operator) => searchActions.updateLogicOperator(clauseId, value),
    handleRemoveClause: (clauseId: string) => searchActions.removeClause(clauseId),
    dispatchAddClause: () => searchActions.addClause(),
    dispatchfetchEvidencesIfNeeded: evidencesType => searchActions.fetchEvidencesIfNeeded(evidencesType),
    dispatchFetchSearchTermsIfNeeded: () => searchActions.fetchSearchTermsIfNeeded(),
    dispatchSubmitAdvancedQuery: () => searchActions.submitAdvancedQuery(),
    dispatchUpdateClauses: clauses => searchActions.updateClauses(clauses),
    dispatchUpdateQueryString: queryString => searchActions.updateQueryString(queryString),
  },
  dispatch,
);

const SearchContainer = withRouter(
  connect(
    mapStateToProps,
    mapDispatchToProps,
  )(Search),
);

export default SearchContainer;
