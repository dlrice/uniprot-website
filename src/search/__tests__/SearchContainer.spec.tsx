import React from 'react';
import '@babel/polyfill';
import { shallow, configure } from 'enzyme';
import Adapter from 'enzyme-adapter-react-16';
import { createEmptyClause } from '../utils/clause';
import { Search } from '../SearchContainer';

configure({ adapter: new Adapter() });

let wrapper;
let props;

describe('Search shallow components', () => {
  beforeEach(() => {
    props = {
      dispatchAddClause: jest.fn(),
      dispatchSubmitQuery: jest.fn(),
      dispatchFetchSearchTerms: jest.fn(),
      dispatchfetchEvidencesIfNeeded: jest.fn(),
      dispatchCopyQueryClausesToSearch: jest.fn(),
      history: {
        push: jest.fn(),
      },
      clauses: [...Array(4)].map(() => createEmptyClause()),
      namespace: 'UniProtKB',
      searchTerms: {
        data: [],
      },
      evidences: {
        go: {
          data: [],
          isFetching: false,
        },
        annotation: {
          data: [],
          isFetching: false,
        },
      },
    };
    wrapper = shallow(<Search {...props} />);
  });

  test('should render', () => {
    expect(wrapper.debug()).toMatchSnapshot();
  });
});
