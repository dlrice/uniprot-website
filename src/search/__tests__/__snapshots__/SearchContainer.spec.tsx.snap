// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Search shallow components should render 1`] = `
ShallowWrapper {
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <Search
    clauses={
      Array [
        Object {
          "id": 1,
          "logicOperator": "AND",
          "queryInput": Object {},
          "searchTerm": Object {
            "dataType": "string",
            "example": "a4_human, P05067, cdc7 human",
            "id": "id_all",
            "itemType": "single",
            "label": "All",
            "term": "All",
          },
        },
        Object {
          "id": 1,
          "logicOperator": "AND",
          "queryInput": Object {},
          "searchTerm": Object {
            "dataType": "string",
            "example": "a4_human, P05067, cdc7 human",
            "id": "id_all",
            "itemType": "single",
            "label": "All",
            "term": "All",
          },
        },
        Object {
          "id": 1,
          "logicOperator": "AND",
          "queryInput": Object {},
          "searchTerm": Object {
            "dataType": "string",
            "example": "a4_human, P05067, cdc7 human",
            "id": "id_all",
            "itemType": "single",
            "label": "All",
            "term": "All",
          },
        },
        Object {
          "id": 1,
          "logicOperator": "AND",
          "queryInput": Object {},
          "searchTerm": Object {
            "dataType": "string",
            "example": "a4_human, P05067, cdc7 human",
            "id": "id_all",
            "itemType": "single",
            "label": "All",
            "term": "All",
          },
        },
      ]
    }
    dispatchAddClause={[MockFunction]}
    dispatchCopyQueryClausesToSearch={[MockFunction]}
    dispatchFetchSearchTerms={[MockFunction]}
    dispatchSubmitQuery={[MockFunction]}
    dispatchfetchEvidencesIfNeeded={[MockFunction]}
    evidences={
      Object {
        "annotation": Object {
          "data": Array [],
          "isFetching": false,
        },
        "go": Object {
          "data": Array [],
          "isFetching": false,
        },
      }
    }
    history={
      Object {
        "push": [MockFunction],
        "replace": [MockFunction],
      }
    }
    location={
      Object {
        "search": "",
      }
    }
    namespace="UniProtKB"
    searchTerms={
      Object {
        "data": Array [],
      }
    }
  />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateError": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "function",
    "props": Object {
      "children": <Unknown
        onChange={[Function]}
        onSubmit={[Function]}
        searchTerm=""
      />,
    },
    "ref": null,
    "rendered": Object {
      "instance": null,
      "key": undefined,
      "nodeType": "function",
      "props": Object {
        "onChange": [Function],
        "onSubmit": [Function],
        "searchTerm": "",
      },
      "ref": null,
      "rendered": null,
      "type": [Function],
    },
    "type": Symbol(react.fragment),
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "function",
      "props": Object {
        "children": <Unknown
          onChange={[Function]}
          onSubmit={[Function]}
          searchTerm=""
        />,
      },
      "ref": null,
      "rendered": Object {
        "instance": null,
        "key": undefined,
        "nodeType": "function",
        "props": Object {
          "onChange": [Function],
          "onSubmit": [Function],
          "searchTerm": "",
        },
        "ref": null,
        "rendered": null,
        "type": [Function],
      },
      "type": Symbol(react.fragment),
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
        "lifecycles": Object {
          "componentDidUpdate": Object {
            "onSetState": true,
          },
          "getDerivedStateFromProps": true,
          "getSnapshotBeforeUpdate": true,
          "setState": Object {
            "skipsComponentDidUpdateOnNullish": true,
          },
        },
      },
    },
  },
}
`;
